neo/core/spiketrainlist.py

```diff
@@ -0,0 +1,150 @@
+        if self._items is None:
+            if self._spike_time_array is None:
+                return str([])
+            else:
+                return "SpikeTrainList containing {} spikes from {} neurons".format(
+                    self._spike_time_array.size,
+                    self._channel_id_array.size)
+        else:
+            return str(self._items)
+
+    def __len__(self):
+        """Return len(self)"""
+        if self._items is None:
+            if self._all_channel_ids is not None:
+                return len(self._all_channel_ids)
+            elif self._channel_id_array is not None:
+                return np.unique(self._channel_id_array).size
+            else:
+                return 0
+        else:
+            return len(self._items)
+
+    def __add__(self, other):
+        """Return self + other"""
+        if isinstance(other, self.__class__):
+            if self._items is None or other._items is None:
+                # todo: update self._spike_time_array, etc.
+                raise NotImplementedError
+            else:
+                self._items.extend(other._items)
+            return self
+        elif other and isinstance(other[0], SpikeTrain):
+            for obj in other:
+                obj.segment = self.segment
+            self._items.extend(other)
+            return self
+        else:
+            return self._items + other
+
+    def __radd__(self, other):
+        """Return other + self"""
+        if self._items is None:
+            self._spiketrains_from_array()
+        other.extend(self._items)
+        return other
+
+    def append(self, obj):
+        """L.append(object) -> None -- append object to end"""
+        if not isinstance(obj, SpikeTrain):
+            raise ValueError("Can only append SpikeTrain objects")
+        if self._items is None:
+            self._spiketrains_from_array()
+        obj.segment = self.segment
+        self._items.append(obj)
+
+    def extend(self, iterable):
+        """L.extend(iterable) -> None -- extend list by appending elements from the iterable"""
+        if self._items is None:
+            self._spiketrains_from_array()
+        for obj in iterable:
+            obj.segment = self.segment
+        self._items.extend(iterable)
+
+    @classmethod
+    def from_spike_time_array(cls, spike_time_array, channel_id_array,
+                              all_channel_ids=None, units='ms',
+                              t_start=None, t_stop=None):
+        """Create a SpikeTrainList object from an array of spike times
+        and an array of channel ids."""
+        obj = cls()
+        obj._spike_time_array = spike_time_array
+        obj._channel_id_array = channel_id_array
+        obj._all_channel_ids = all_channel_ids
+        obj._spiketrain_metadata = {
+            "units": units,
+            "t_start": t_start,
+            "t_stop": t_stop
+        }
+        return obj
+
+    def _spiketrains_from_array(self):
```

Comment:
This function is quite expensive, as it generates all Spiketrain objects explicitly. As it is used in many of Spiketrainlist methods, this would cause quite a bit of overhead. Maybe it would be good to use it in only as few places as possible?

Reply:
It's only ever used once, because the generated SpikeTrain objects are cached in `self._items`

Reply:
Yes, you are right. So in this PR the `SpikeTrain` representation is still the reference all data will be converted to eventually. Even when merging two `SpikeTrainList`s coming from a spike_time_array representation. Is there any way we can avoid that? Maybe having a default operation mode that can switch between `SpikeTrain` and `spike_time_list`? Or would it be worth to have the `_spike_time_array` and the corresponding masks for individual channels / units as base representation and only generate SpikeTrains when required (without) caching / duplicating the data? This would rely even stronger on numpy for inserting / reordering of the spike_times lists and masks. I am not sure which method would be more performant in the end. @apdavison  Do you have a better intuition there?

Reply:
you're right that merging two array-representation SpikeTrainLists should avoid creating `SpikeTrain`s. I think that's easy enough to do. 

The general idea is to keep data in the representation they arrive in as long as possible, to avoid unecessary transformations, i.e. the "reference/base" representation depends on how the object was initialized. 

Note that the `multiplexed` property (not yet implemented in this branch, but you can see it in the [spiketrainlist branch](https://github.com/apdavison/python-neo/blob/spiketrainlist/neo/core/spiketrainlist.py#L170) which will replace this PR) allows users to access the array representation.